### ISSUE TITLE
reducer callback from_argument: x or data

### ISSUE THREAD

**Issue by przell** [Open-EO/openeo-r-client#47]
Hi Florian, 
I just tried to to do a temporal reduce with a mean as the reducer. 
In the Reducer -> callback -> data -> from_argument the r-client gives "x". On the webeditor "data" is inserted here by default. @prateekbudhwar also checked on the python client and it gives "data" as well. 
Here is a snippet from the r code I used:
```
vv_mean = p$reduce(data = vv, dimension = "temporal", reducer = p$mean)
```
Here is the process graph produced by the r-client:
```
  "reduce_TWBUX5887C": {
    "process_id": "reduce",
    "arguments": {
      "data": {
        "from_node": "filter_bands_GMGTA1945Y"
      },
      "reducer": {
        "callback": {
          "mean_OASAU8940N": {
            "result": true,
            "process_id": "mean",
            "arguments": {
              "data": {
                "from_argument": "x"
              },
              "ignore_nodata": {
                "from_argument": "y"
              }
            }
          }
        }
      },
      "dimension": "temporal"
    }
  }
```
And the one from the webeditor:
```
  "2": {
    "process_id": "reduce",
    "arguments": {
      "data": {
        "from_node": "3"
      },
      "reducer": {
        "callback": {
          "2": {
            "result": true,
            "process_id": "mean",
            "arguments": {
              "data": {
                "from_argument": "data"
              }
            }
          }
        }
      },
      "dimension": "temporal"
    }
  }
```
Changing the `from_argument` to "data" in the graph produced by the r client works in the webeditor! 
The r-client also inserts the `ignore_nodata` part. This doesn't break the graph. But it is not present on the web editor. Is this intended?
I'm using `openeo 0.6.1` (dev)

@prateekbudhwar thanks for checking up on this!

**Comment by flahn** [comment]
This is tricky indeed and it is somewhat intended. `ignore_nodata` is specified by your process. This is conform to the openeo processes description. 

```R
process_viewer(x = "reduce")
```

In RStudio this opens a viewer showing the EURAC specification of the given process. Currently we have two options on callback parameter - a unary choice of `data` and a binary choice with `x` and `y`. When I allow a function or a process as a callback, I assume which of those two choices is mostly appropriate for the callback. I do this by matching the number of arguments rather than the correct names, because they differed on different backends. The main reason was to split binary and unary operators. Because `mean` offers two parameters `data` and `ignore_nodata` the choice is made to use `x` and `y`.

Back, to the problem and a solution: You can use a specific function to control the behavior.

```r
vv_mean = p$reduce(data = s2, dimension = "temporal", reducer = function(x) {
  p$mean(data = x)
})
```

We will have no overloaded reducer anymore in the openEO process 1.0.0 RC1. In this cases we will have two separate function `reduce_dimension` and `reduce_dimension_binary`.

**Comment by przell** [comment]
Hi Florian, 
thanks for the precise answer. I see in the process graph, that this solution works.
But: Now the argument "dimension" is not added anymore to the process graph.
```
vv_mean = p$reduce(data = vv, dimension = "temporal", reducer = function(x) {
  p$mean(data = x)
})
```

```
  "reduce_SZIPH5191Y": {
    "process_id": "reduce",
    "arguments": {
      "target_dimension": null,
      "data": {
        "from_node": "filter_bands_QGZID9826O"
      },
      "binary": false,
      "reducer": {
        "callback": {
          "mean_LZWXH2806S": {
            "process_id": "mean",
            "arguments": {
              "data": {
                "from_argument": "data"
              },
              "ignore_nodata": true
            },
            "result": true
          }
        }
      }
    }
  }
```
I'm using openeo r-client dev 0.6.1.
And am connected to eurac backend (https://openeo.eurac.edu) api version 0.4.2
When I copy the process graph into the webeditor and add "dimension" manually it works.
When I run the process graph through the rclient I receive this error:
```
SERVER-ERROR: <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>Apache Tomcat/7.0.76 - Error report</title>
<style><!--H1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} H2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} H3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} BODY {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} B {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} P {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;}A {color : black;}A.name {color : black;}HR {color : #525D76;}--></style>
</head>
<body>
<h1>HTTP Status 500 - org.json.JSONException: JSONObject["dimension"] not found.</h1>
<hr size="1" noshade>
<p><b>type</b> Exception report</p>
<p><b>message</b> <u>org.json.JSONException: JSONObject["dimension"] not found.</u></p>
<p><b>description</b> <u>The server encountered an internal error that prevented it from fulfilling this request.</u></p>
<p><b>exception</b> </p>
<pre>javax.servlet.ServletException: org.json.JSONException: JSONObject["dimension"] not found.
	org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:489)
	org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
	org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
	org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
	org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	eu.openeo.api.ApiOriginFilter.doFilter(ApiOriginFilter.java:42)
</pre>
<p><b>root cause</b> </p>
<pre>org.json.JSONException: JSONObject["dimension"] not found.
	org.json.JSONObject.get(JSONObject.java:472)
	org.json.JSONObject.getString(JSONObject.java:718)
	eu.openeo.backend.wcps.WCPSQueryFactory.executeProcesses(WCPSQueryFactory.java:3253)
	eu.openeo.backend.wcps.WCPSQueryFactory.parseOpenEOProcessGraph(WCPSQueryFactory.java:3180)
	eu.openeo.backend.wcps.WCPSQueryFactory.build(WCPSQueryFactory.java:96)
	eu.openeo.backend.wcps.WCPSQueryFactory.&lt;init&gt;(WCPSQueryFactory.java:56)
	eu.openeo.api.impl.ResultApiServiceImpl.resultPost(ResultApiServiceImpl.java:32)
	eu.openeo.api.ResultApi.resultPost(ResultApi.java:66)
	sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	java.lang.reflect.Method.invoke(Method.java:498)
	org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
	org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
	org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
	org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160)
	org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
	org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
	org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
	org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
	org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
	org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
	org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
	org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
	org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
	org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
	org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
	org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
	org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	eu.openeo.api.ApiOriginFilter.doFilter(ApiOriginFilter.java:42)
</pre>
<p><b>note</b> <u>The full stack trace of the root cause is available in the Apache Tomcat/7.0.76 logs.</u></p>
<hr size="1" noshade>
<h3>Apache Tomcat/7.0.76</h3>
</body>
</html>
```

**Comment by przell** [comment]
Thanks @flahn! Your fix works!
If you think that the issue can be closed please go ahead!
